### PR TITLE
[SPARK-46202][CONNECT] Expose new ArtifactManager APIs to support custom target directories

### PR DESCRIPTION
**Pull request by vicennial**
### What changes were proposed in this pull request?
Adds new client APIs to the Spark Connect Scala Client:
- `def addArtifact(bytes: Array[Byte], target: String): Unit`
- `def addArtifact(source: String, target: String): Unit`

### Why are the changes needed?
Currently, without the use of a REPL/Class finder, there is no API to support adding artifacts (file-based and in-memory) 
 with a custom target directory structure to the remote Spark Connect session. 

### Does this PR introduce _any_ user-facing change?
Yes.

Users can do the following for classfiles and jars:
```scala
addArtifact("/Users/dummyUser/files/foo/bar.class", "sub/directory/foo.class")
addArtifact(bytesBar, "bar.class")
```

This would preserve the directory structure in the remote server. In this case, the file would be stored under the directory:
`$ROOT_SESSION_ARTIFACT_DIRECTORY/classes/bar.class`
`$ROOT_SESSION_ARTIFACT_DIRECTORY/classes/sub/directory/foo.class`
### How was this patch tested?

Unit test

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala [hunk 1]
```diff
@@ -408,14 +432,28 @@ object Artifact {
     jars.map(p => Paths.get(p)).map(path => newJarArtifact(path.getFileName, new LocalFile(path)))
   }
 
+  private def concatenatePaths(basePath: Path, otherPath: Path): Path = {
+    // We avoid using the `.resolve()` method here to ensure that we're concatenating the two
+    // paths even if `otherPath` is absolute.
+    val concatenatedPath = Paths.get(basePath.toString + "/" + otherPath.toString)
```
Comment (HyukjinKwon) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala:438]:
```suggestion
    val concatenatedPath = Paths.get(basePath.toString, otherPath.toString)
```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -590,6 +590,16 @@ class SparkSession private[sql] (
   @Experimental
   def addArtifact(uri: URI): Unit = client.addArtifact(uri)
 
+  /**
+   * Add a single in-memory artifact to the client session with a specified target URI.
+   *
+   * Currently it supports files with extensions .jar and .class.
```
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:596]: In API documentation, we should better remove things like "Currently".  Should say sth like "Supported file extensions are `.jar` and `.class`".

**Comment by vicennial** [comment]
cc @hvanhovell @nija-at

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -590,6 +590,48 @@ class SparkSession private[sql] (
...
+   *  // ${WORKING_DIR_FOR_CLASS_FILES}/foo/bar.class
+   * }}}
+   *
+   * @since 4.0.0
+   */
+  @Experimental
+  def addArtifact(bytes: Array[Byte], target: String): Unit = client.addArtifact(bytes, target)
+
+  /**
+   * Add a single in-memory artifact to the session while preserving the directory structure
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:615]: Can you update the docs?

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala [hunk 4]
```diff
@@ -108,6 +104,56 @@ class ArtifactManager(
...
+   * ==Example==
+   * {{{
+   *  addArtifact(bytesBar, "foo/bar.class")
+   *  addArtifact(bytesFlat, "flat.class")
+   *  // Directory structure of the session's working directory for class files would look like:
+   *  // ${WORKING_DIR_FOR_CLASS_FILES}/flat.class
+   *  // ${WORKING_DIR_FOR_CLASS_FILES}/foo/bar.class
+   * }}}
+   */
+  def addArtifact(bytes: Array[Byte], target: String): Unit = {
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala:123]: Where is this used? I am not against it, just trying to understand?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 3 above)
Comment (vicennial) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala]: Oops, fixed, good eye!

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala (hunk 4 above)
Comment (vicennial) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala:123]: It's not used at the moment.  We do, however, use a similar variation of this (without a custom `target`)  in the `AmmoniteClassFinder`.  This API would let user's do the same if they wish to.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Pull request closed without merging**